services/admin/src/store.ts

```diff
@@ -277,16 +285,55 @@ export class Store {
 
         const lines = readline.createInterface(fs.createReadStream(filePath));
 
+        const currentCvrFileMode =
+          this.getCurrentCvrFileModeForElection(electionId);
+        let lastLineCvrMode: Optional<Admin.CvrFileMode>;
+        let nextLineCvrMode: Optional<Admin.CvrFileMode>;
+
         for await (const line of lines) {
           if (line.trim() === '') {
             continue;
           }
 
-          // TODO(https://github.com/votingworks/vxsuite/issues/2716): add error checking against
-          // attempts to mix "live" and "test" CVRs for the same election.
+          // Parse CVR data for validation.
+          const cvr = safeParseJson(line).unsafeUnwrap() as CastVoteRecord;
```

Comment:
If we're parsing here, should we use the parsed data in later functions? I think I see that `addCastVoteRecordFileEntry` also parses the line, for one. Not sure if there are others. Maybe not an important optimization, but for large CVR files it might make a difference.

Reply:
ah thanks, agreed - had made that change at some point and can't remember why i reverted!

Reply:
ah yeah, i realised that the sub-method would have to re-stringify the CVR for storage anyway which is why i reverted earlier - but, with some refactoring, was able to avoid the double parsing and it's hopefully looking a little cleaner now